### PR TITLE
[Reviewer: Adam] Fix timer store uts

### PR DESCRIPTION
**Pull request by sebrexmetaswitch**
As discussed, I'm probably going to get Ellie to review this too, but wanted someone in our team to review it first to build up expertise etc.

This change was motivated by https://github.com/Metaswitch/clearwater-issues/issues/2489, where we occasionally missed coverage depending on when the UTs were run.

However, it expanded a bit (sticking to our "make things a bit better" philosophy) to include some tidy up of the `TimerStore` and other minor tweaks.

`make`, `make full_test` and `make fv_test` all pass. I haven't yet run the live tests over it, but I waned to get eyes on it first. I'm probably just looking to merge this into master, so there's no huge rush on it.

### Changes
I made a few changes in sections, so I'd recommend reviewing the commits I've pulled out here separately, as it'll all make more sense that way.

1. Fix up the unreliable UT
    * It turns out that we didn't always hit this line in the UTs: https://github.com/Metaswitch/chronos/blob/db26d04bc8bfdafa7686d847884c05fa38e1c483/src/timer_store.cpp#L432. Having looked at the test sometimes hit it, it was the `TestTimerHandlerRealStore` `TimeFromShortWheelTimers` test in `test_timer_store.cpp`. That tests adds 2 timers to the short wheel, 10ms apart. We missed coverage if we happened to control time at a point such that these 2 timers went into different buckets in the short wheel, and the time_from fell into the 2nd of the buckets. In that case, when we retreived the timers, we started from the second bucket so we never had to skip over the first one. I verified this was the case by editing the test to advance time to one such time, and it failed each time.
    * To fix this, I wanted to ste timers carefully such that they were always in the same bucket. However, this sort of test requires knowledge of the internals of the `TimerStore`, so I added this test in `test_timer_store.cpp` not `test_timer_handler.cpp` (The handler tests should not care about how the store organises its data).
    * I added this UT in the first commit (https://github.com/Metaswitch/chronos/commit/a855ba5ed373f5c5b7c2b3a359b4229c2b5a17df). It basically ensures that it puts both timers into the same bucket in the short wheel.

1. In adding this UT, I found another bug, which caused my new UT to fail.
    * The bug was that if the `time_from` fell into a bucket which was "earlier" that the current time, we didn't set the `_end_bucket` correctly. e.g. if a wheel had 4 buckets, and the current time was in bucket 3 and `time_from` was in bucket 0, then we'd end up starting at bucket 0 and iterating until we hit bucket 7. Since the bucket indices are modulo 4, this meant we'd go through buckets 0, 1 and 2 twice.
    * I fixed this in the second commit (https://github.com/Metaswitch/chronos/commit/ec40c8fafd752854df6e8a6ed4b450937ebc574c). There's a fairly detailed comment in that commit, so I won't reproduce it here.
    * The UT I added in the first commit was testing this, so I didn't add any others for it.

1. Because we had separate long and short wheel iterators, it followed that the long wheel had the same bug. So I added a UT to verify that it did (https://github.com/Metaswitch/chronos/commit/4241a8b17d26be0aaaa3478f1f9837fa9c4d8ef5) which failed until I also fixed the bug for the long wheel (https://github.com/Metaswitch/chronos/commit/a284346ffd02fae6b147830300492e8de575c16d)

1. You'll have noticed that the short and long wheel iterators now have a whole load of duplicated code, so I wanted to refactor them. I wanted to do this after I had put in place all the UTs and got them passing, so I could verify that the small refactor didn't break things, which is why I did this separately.
    * I created a new `TSBaseWheelIterator` which both the `TSShortWheelIterator` and the `TSLongWheelIterator` inherit from.
    * I then commonalised all the methods from the two previous iterators into the base one, because they're fundamentally doing the same thing
    * I added virtual methods for the few things that are different (refilling the wheel, getting a bucket and rouding to the wheel resolution)
    * the last 4 commits (https://github.com/Metaswitch/chronos/pull/382/files/a284346ffd02fae6b147830300492e8de575c16d..7a2c799f84796d2da9fc385188e1b2c294da56d1) are this refactor

### CODE REVIEW

File: src/ut/test_timer_store.cpp [hunk 1]
```diff
@@ -684,6 +684,43 @@ TYPED_TEST(TestTimerStore, IterateOverLongWheelTimersTimeFrom)
   delete timer5;
 }
 
+// Test that the BaseWheelIterator will correctly skip all timers in a bucket if
+// they're earlier than the time we're iterating from.
+TYPED_TEST(TestTimerStore, IterateOverTimersSkipEarlier)
+{
+  // We want to add one timer to the next short wheel bucket, and one to the
+  // one after that, to test that we correct skip over the timer in the next
```
Comment (johadalin) [inline src/ut/test_timer_store.cpp]: `correctly`

**Review by johadalin — COMMENTED** [review]
Generally i think this looks good. Like the continued application of 'leave things better than you found them', and properly fixing it all up. 
I've got a number of comments, mostly on your comments.  I definitely think it's worth Ellie giving it a proper review as well though, as i don't think i'm up enough on how the chronos internals work to give a view on whether this will have any unintended effects or anything like that.

**Comment by sebrexmetaswitch** [comment]
I actually think that I may have missed the case when `time_from` is behind the current time. In that case, we want to make sure that the iterator starts from the current bucket, not the `time_from` bucket

File: include/timer_store.h [hunk 2]
```diff
@@ -91,32 +103,85 @@ class TimerStore
...
   private:
+    const int _resolution;
+    const int _num_buckets;
+    const int _period;
     int _end_bucket;
     int _bucket;
     void next_bucket();
+
+    /// Kick the timer store to refill this wheel.
+    virtual void refill_wheel_from_timer_store() = 0;
```
[the remark this answers is not in the document]
Comment (sebrexmetaswitch) [reply include/timer_store.h]: Yeah, but I was putting comments on all the other methods and I didn't want this one to feel left out. I was hoping I'd have more to say

**Comment by sebrexmetaswitch** [comment]
@eleanor-merry As discussed, I'd like your opinion on these changes too. Happy to wait a week or two.

**Review by eleanor-merry — APPROVED** [review]
LGTM

**Pull request closed without merging**